### PR TITLE
Add PHP 7.4, fix Psalm issues

### PR DESCRIPTION
**Pull request by Slamdunk**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -27,7 +27,7 @@
   },
   "require-dev": {
     "phpunit/phpunit": "^6|^7",
-    "vimeo/psalm": "^1|^2"
+    "vimeo/psalm": "^2|^3"
```
Comment (Slamdunk) [inline composer.json:30]: Version 2 has 7.0 as minimum requirement, so v1 is not required anymore

File: psalm.xml [hunk 2]
```diff
@@ -10,5 +10,10 @@
     <issueHandlers>
         <LessSpecificReturnStatement errorLevel="info" />
         <MoreSpecificReturnType errorLevel="info" />
+        <InternalMethod>
+            <errorLevel type="suppress">
+                <referencedMethod name="ParagonIE_Sodium_Core_Util::store64_le" />
```
Comment (Slamdunk) [inline psalm.xml:15]: See https://github.com/paragonie/paseto/issues/111 point [1]: IMHO `ParagonIE_Sodium_Core_Util` should be a isolated package

File: src/Protocol/Version2.php [hunk 3]
```diff
@@ -158,12 +158,17 @@ public static function decrypt(
...
         }
-        return self::aeadDecrypt(
+        $message = self::aeadDecrypt(
             $data,
             self::HEADER . '.local.',
             $key,
             (string) $footer
         );
+        if (!\is_string($message)) {
+            throw new PasetoException('Invalid message decryption');
```
Comment (Slamdunk) [inline src/Protocol/Version2.php:168]: Cannot test this, but needed for proper return type consistency

File: src/Protocol/Version2.php [hunk 4]
```diff
@@ -318,7 +325,7 @@ public static function aeadDecrypt(
         string $header,
         SymmetricKey $key,
         string $footer = ''
-    ): string {
+    ) {
```
Comment (Slamdunk) [inline src/Protocol/Version2.php:328]: The extension function may return false too:

https://github.com/jedisct1/libsodium-php/blob/28d13bf437cb969a0583031fc7ac54c5a8dc8116/libsodium.c#L3068

File: src/ProtocolCollection.php [hunk 5]
```diff
@@ -96,7 +96,6 @@ public static function throwIfUnsupported(ProtocolInterface $protocol)
      */
     public static function protocolFromHeaderPart(string $headerPart): ProtocolInterface {
         if (empty(self::$headerLookup)) {
-            /** @var ProtocolInterface $protocolClass */
```
Comment (Slamdunk) [inline src/ProtocolCollection.php:99]: This messes IDEs, but Psalm correctly infer type from `@const array<int, class-string<ProtocolInterface>>` notation added above

**Comment by Slamdunk** [comment]
Mmm, Travis build wasn't triggered:

immagine

File: .travis.yml [hunk 6]
```diff
@@ -1,10 +1,14 @@
...
-- 7.3
+matrix:
+  fast_finish: true
+  include:
+    - php: 7.0
+    - php: 7.1
+    - php: 7.2
+    - php: 7.3
+    - php: 7.4
+      env: FULL_TEST=1
```
Comment (Slamdunk) [inline .travis.yml:11]: The error given in the last failing build:
https://travis-ci.org/github/paragonie/paseto/jobs/677231516#L318-L321
is due to Psalm bugs fixed in more recent version, but not in the older ones.
In general, test shall be done on every version, of course, but static analysis is enough only on the latest one.

**Comment by Slamdunk** [comment]
Ping @paragonie-security

**Comment by lhchavez** [comment]
would you mind also adding

```diff
diff --git a/src/Parsing/PasetoMessage.php b/src/Parsing/PasetoMessage.php
index 8fe4839..132775d 100644
--- a/src/Parsing/PasetoMessage.php
+++ b/src/Parsing/PasetoMessage.php
@@ -47,7 +47,6 @@ final class PasetoMessage
      * @throws SecurityException
      * @throws InvalidVersionException
      * @throws InvalidPurposeException
-     * @return PasetoMessage
      * @throws \TypeError
      */
     public static function fromString(string $tainted): self
```

? That fixes another issue with the latest Psalm, since that function has a duplicate `@return` docstring tag.

**Comment by Slamdunk** [comment]
@lhchavez done, thank you

**Comment by Slamdunk** [comment]
ping @paragonie-scott @paragonie-security @paragonie-staff

File: src/Protocol/Version2.php (hunk 4 above)
Comment (paragonie-security) [reply src/Protocol/Version2.php:328]: If that happens, then we need to be stricter.

Comment (Slamdunk) [reply src/Protocol/Version2.php:328]: Isn't https://github.com/paragonie/paseto/pull/113/files#diff-06d71742cf83f525040bc36171832f85796debb414dbae04c7c1132254381ec5R167 enough?

**Comment by Slamdunk** [comment]
@paragonie-security as soon as this gets merged, I'll add a PR for PHP 8 compatibility

**Comment by stayallive** [comment]
I was looking to see if PHP 8 was on the radar, would be greatly appreciated if that is happening 👍

**Pull request merged**